### PR TITLE
Variable size singularities

### PR DESCRIPTION
**Pull request by cogreaper**
[FEATURE] [BALANCE]

## About the PR 
Allows singularities to spawn in odd square sizes up to 11x11 instead of only allowing 5x5. power scales with the size of the singularity.

## Why's this needed? 
You can hellburn for immense power. You can stack hotspots for immense power. You cant do anything comparable with a singularity. This fixes that. Side note, giant singularities are fun, especially for antags or admins(whats the difference even).

## Changelog

```changelog
(u)Anachroniser:
(*)Singularities now come in variable sizes and can generate more power as they grow. Keep in mind the risks of a giant singularity.
```

### CODE REVIEW

**Comment by ZeWaka** [comment]
please label your PRs

File: code/obj/machinery/singularity.dm [hunk 1]
```diff
@@ -135,25 +143,19 @@ for some reason I brought it back and tried to clean it up a bit and I regret ev
...
+		if (checkpointC < max(4,(radius*8)))//as radius of a 5x5 should be 2, 16 tiles are needed to hold it in, this allows for 4 failures before the singularity is loose
 			src.active = 1
-			grav_pull = 8
 
-/obj/machinery/the_singularity/emp_act()
-	return // No action required this should be the one doing the EMPing
 
 /obj/machinery/the_singularity/proc/eat()
-	for (var/X in range(grav_pull, src.get_center()))
+	for (var/X in orange(grav_pull, src.get_center()))
```
Comment (Azrun) [inline code/obj/machinery/singularity.dm:155]: Please test that contents of containers are consumed.

File: code/obj/machinery/singularity.dm [hunk 2]
```diff
@@ -374,7 +366,7 @@ for some reason I brought it back and tried to clean it up a bit and I regret ev
 	pulse.icon_state = "emppulse"
 	pulse.name = "emp pulse"
 	pulse.anchored = 1
-	SPAWN_DBG(2 SECONDS)
+	SPAWN_DBG (20)
```
Comment (Azrun) [inline code/obj/machinery/singularity.dm:369]: Please use readable time instead of deciseconds.

File: code/obj/machinery/singularity.dm [hunk 3]
```diff
@@ -554,7 +551,7 @@ for some reason I brought it back and tried to clean it up a bit and I regret ev
 		T2 = T
 		var/obj/machinery/containment_field/CF = new/obj/machinery/containment_field/(src, G) //(ref to this gen, ref to connected gen)
 		CF.set_loc(T)
-		CF.set_dir(field_dir)
+		CF.dir = field_dir
```
Comment (Azrun) [inline code/obj/machinery/singularity.dm:554]: There was an effort to use set_dir over other operations, this resolve something that helper does not?

File: code/obj/machinery/singularity.dm [hunk 4]
```diff
@@ -1389,31 +1385,31 @@ for some reason I brought it back and tried to clean it up a bit and I regret ev
...
 			var/power_p = 0
 
 			if(!isnull(S1))
-				power_s += S1.energy
+				power_s += S1.energy*max((S1.radius**2),1)/4
 			if(P1?.air_contents)
 				if(CA1.active != 0)
 					power_p += P1.air_contents.toxins
-					P1.air_contents.toxins -= 0.001 * mult
+					P1.air_contents.toxins -= 0.001
```
Comment (ZeWaka) [inline code/obj/machinery/singularity.dm:1400]: why tf did you remove mult

File: code/obj/machinery/singularity.dm [hunk 5]
```diff
@@ -1295,14 +1296,9 @@ for some reason I brought it back and tried to clean it up a bit and I regret ev
 
 /obj/machinery/power/collector_control/New()
 	..()
-	START_TRACKING
 	SPAWN_DBG(1 SECOND)
 		updatecons()
 
-/obj/machinery/power/collector_control/disposing()
-	. = ..()
-	STOP_TRACKING
```
Comment (ZeWaka) [inline code/obj/machinery/singularity.dm:1304]: why do this

File: code/obj/machinery/singularity.dm [hunk 6]
```diff
@@ -1209,14 +1206,16 @@ for some reason I brought it back and tried to clean it up a bit and I regret ev
...
+			CU.updatecons()
 		boutput(user, "You turn off the collector array.")
 		return
 
 	if(src.active==0)
 		src.active = 1
 		icon_state = "ca_active"
-		CU?.updatecons()
+		if(CU)
+			CU.updatecons()
```
Comment (ZeWaka) [inline code/obj/machinery/singularity.dm:1218]: WHY

File: code/obj/machinery/singularity.dm [hunk 7]
```diff
@@ -1714,11 +1710,10 @@ for some reason I brought it back and tried to clean it up a bit and I regret ev
 	else
 		var/seconds = src.time % 60
 		var/minutes = (src.time - seconds) / 60
-		var/flick_seperator = (seconds % 2 == 0)  || !src.timing
 		minutes = minutes < 10 ? "0[minutes]" : "[minutes]"
 		seconds = seconds < 10 ? "0[seconds]" : "[seconds]"
 
-		return "[minutes][flick_seperator ? ":" : " "][seconds]"
+		return "[minutes][seconds % 2 == 0 ? ":" : " "][seconds]"
```
Comment (ZeWaka) [inline code/obj/machinery/singularity.dm:1716]: why?????

File: code/obj/machinery/singularity.dm [hunk 8]
```diff
@@ -1423,7 +1419,7 @@ for some reason I brought it back and tried to clean it up a bit and I regret ev
 		var/power_s = 0
 		var/power_p = 0
 		if(!isnull(S1))
-			power_s += S1.energy
+			power_s += S1.energy*((S1.radius*2+1)**2)/25//should give the area of the singularity and divide it by the area of a standard singularity(a 5x5)
```
Comment (ZeWaka) [inline code/obj/machinery/singularity.dm:1422]: should use defines for this

File: code/obj/machinery/singularity.dm [hunk 9]
```diff
@@ -996,7 +993,7 @@ for some reason I brought it back and tried to clean it up a bit and I regret ev
 /obj/machinery/emitter/attackby(obj/item/W, mob/user)
 	if (ispryingtool(W))
 		if(!anchored)
-			src.set_dir(turn(src.dir, -90))
+			src.dir = turn(src.dir, -90)
```
Comment (ZeWaka) [inline code/obj/machinery/singularity.dm:996]: no

File: code/obj/machinery/singularity.dm [hunk 10]
```diff
@@ -821,7 +818,7 @@ for some reason I brought it back and tried to clean it up a bit and I regret ev
 	// Added (Convair880).
 	logTheThing("combat", user, null, "was shocked by a containment field at [log_loc(src)].")
 
-	if (user?.bioHolder)
+	if (user && user.bioHolder)
```
Comment (ZeWaka) [inline code/obj/machinery/singularity.dm:821]: WHY

File: code/obj/machinery/singularity.dm [hunk 11]
```diff
@@ -527,13 +520,17 @@ for some reason I brought it back and tried to clean it up a bit and I regret ev
 	else if(NSEW == 8)
 		oNSEW = 4
 
-	for(var/dist = 0, dist <= 9, dist += 1) // checks out to 8 tiles away for another generator
+	for(var/dist = 0, dist <= 11, dist += 1) // checks out to 11(hopefully) tiles away for another generator
```
Comment (ZeWaka) [inline code/obj/machinery/singularity.dm:523]: what does hopefully mean

File: code/obj/machinery/singularity.dm [hunk 12]
```diff
@@ -270,10 +262,10 @@ for some reason I brought it back and tried to clean it up a bit and I regret ev
 	src.energy += gain
 
 /obj/machinery/the_singularity/proc/get_center()
-	. = get_turf(src)
+	//. = get_turf(src)
 	//if(!(get_step(., SOUTHWEST) in src.locs)) // I hate this, neither `loc` nor `get_turf` behave consistently, sometimes they are the center tile and sometimes they are the south west tile, aaaa
-	if(has_moved)
-		. = get_step(., NORTHEAST)
+	return loc
```
Comment (ZeWaka) [inline code/obj/machinery/singularity.dm:267]: umm why

File: code/obj/machinery/singularity.dm [hunk 13]
```diff
@@ -205,13 +203,8 @@ for some reason I brought it back and tried to clean it up a bit and I regret ev
 		if (A.event_handler_flags & IMMUNE_SINGULARITY_INACTIVE)
 			return
 
-	// Don't bump that which no longer exists
-	if(A.disposed)
-		return
```
Comment (ZeWaka) [inline code/obj/machinery/singularity.dm:210]: why????????????????????

File: code/obj/machinery/singularity.dm [hunk 14]
```diff
@@ -180,21 +182,17 @@ for some reason I brought it back and tried to clean it up a bit and I regret ev
...
 
 /obj/machinery/the_singularity/ex_act(severity, last_touched, power)
-	if(!maxboom)
-		SPAWN_DBG(0.1 SECONDS)
-			if(severity == 1 && (maxboom ? prob(maxboom*5) : prob(30))) //need a big bomb (TTV+ sized), but a big enough bomb will always clear it
-				qdel(src)
-			maxboom = 0
-	maxboom = max(power, maxboom)
+	if(severity == 1 && (power ? prob(power*5) : prob(30))) //need a big bomb (TTV+ sized), but a big enough bomb will always clear it
+		qdel(src)
```
Comment (ZeWaka) [inline code/obj/machinery/singularity.dm:195]: unrelated and why??????????????/

File: code/obj/machinery/singularity.dm [hunk 15]
```diff
@@ -135,25 +143,19 @@ for some reason I brought it back and tried to clean it up a bit and I regret ev
...
-	return // No action required this should be the one doing the EMPing
 
 /obj/machinery/the_singularity/proc/eat()
-	for (var/X in range(grav_pull, src.get_center()))
+	for (var/X in orange(grav_pull, src.get_center()))
 		LAGCHECK(LAG_LOW)
 		if (!X)
 			continue
-		if (X == src)
-			continue
```
Comment (ZeWaka) [inline code/obj/machinery/singularity.dm:155]: no.

File: code/obj/machinery/singularity.dm (hunk 1 above)
Comment (ZeWaka) [inline code/obj/machinery/singularity.dm:155]: why would you change this, this is intentional

File: code/obj/machinery/singularity.dm (hunk 3 above)
Comment (Azrun) [reply code/obj/machinery/singularity.dm:554]: Look through 278c2a7, there are... a number of things I hope you didn't mean to change.

File: code/obj/machinery/singularity.dm [hunk 16]
```diff
@@ -135,25 +143,19 @@ for some reason I brought it back and tried to clean it up a bit and I regret ev
...
 		SPAWN_DBG(1.1 SECONDS) // slowing this baby down a little -drsingh
 			move()
-	else
+	else//this should probably be modified to use the enclosed test of the generator
 		var/checkpointC = 0
-		for (var/obj/machinery/containment_field/X in orange(3,src))
+		for (var/obj/machinery/containment_field/X in orange(radius+2,src))
 			checkpointC ++
-		if (checkpointC < 18)
+		if (checkpointC < max(4,(radius*8)))//as radius of a 5x5 should be 2, 16 tiles are needed to hold it in, this allows for 4 failures before the singularity is loose
```
Comment (ZeWaka) [inline code/obj/machinery/singularity.dm:150]: use defines here

File: code/obj/machinery/singularity.dm [hunk 17]
```diff
@@ -76,35 +83,36 @@ Contains:
...
 	var/energy = 10
-	var/warp = 5
 	var/lastT = 0
-	var/warp_delay = 30
 	var/Dtime = null
 	var/Wtime = 0
 	var/dieot = 0
 	var/selfmove = 1
 	var/grav_pull = 6
+	var/radius = 0 //the variable used for all calculations involving size, please attribute all of the problems this inevitably creates to anachroniser
```
Comment (ZeWaka) [inline code/obj/machinery/singularity.dm:102]: no need for second half of comment

File: code/obj/machinery/singularity.dm [hunk 18]
```diff
@@ -26,11 +26,16 @@ Contains:
...
 */
 /obj/machinery/the_singularitygen/process()
-	var/checkpointC = 0
-	for (var/obj/X in orange(4,src))
-		if (istype(X, /obj/machinery/containment_field))
-			checkpointC ++
-	if (checkpointC >= 20)
+	var/goodgenerators = 0 //ensures that there are 4 generators in place with at least 2 links. note that false positives are very possible and will result in a loose singularity
+	var/smallestdimension = 0//determines the radius of the produced singularity
+	for (var/obj/X in orange(11,src))
```
Comment (ZeWaka) [inline code/obj/machinery/singularity.dm:31]: why the change

**Review by ZeWaka — CHANGES_REQUESTED** [review]
needs massive reworking and thought on why so many unrelated changes were made

File: code/obj/machinery/singularity.dm (hunk 18 above)
Comment (cogreaper) [reply code/obj/machinery/singularity.dm:31]: The change was necessary to establish the size of the singularity that would spawn as that is no longer constant.

Comment (Tarmunora) [reply code/obj/machinery/singularity.dm:31]: `for (var/obj/X in orange(11,src))` 😱 

just... no. Use by_type or something instead

File: code/obj/machinery/singularity.dm (hunk 3 above)
Comment (cogreaper) [reply code/obj/machinery/singularity.dm:554]: Literally every single one of them, *scream

Comment (Azrun) [reply code/obj/machinery/singularity.dm]: 👍

File: code/obj/machinery/singularity.dm [hunk 19]
```diff
@@ -1311,7 +1332,7 @@ for some reason I brought it back and tried to clean it up a bit and I regret ev
 		CAS = locate(/obj/machinery/power/collector_array) in get_step(src,SOUTH)
 		CAE = locate(/obj/machinery/power/collector_array) in get_step(src,EAST)
 		CAW = locate(/obj/machinery/power/collector_array) in get_step(src,WEST)
-		for(var/obj/machinery/the_singularity/S in orange(12,src))
+		for(var/obj/machinery/the_singularity/S in orange(18,src))
```
Comment (ZeWaka) [inline code/obj/machinery/singularity.dm:1335]: should use by_type

File: code/obj/machinery/singularity.dm [hunk 20]
```diff
@@ -334,11 +345,11 @@ for some reason I brought it back and tried to clean it up a bit and I regret ev
 
 /obj/machinery/the_singularity/proc/BHolerip()
 
-	for (var/turf/T in orange(6, src.get_center()))
+	for (var/turf/T in orange(radius*4+4, src.get_center()))
```
Comment (ZeWaka) [inline code/obj/machinery/singularity.dm:348]: why *4+4

File: code/obj/machinery/singularity.dm [hunk 21]
```diff
@@ -334,11 +345,11 @@ for some reason I brought it back and tried to clean it up a bit and I regret ev
...
 		LAGCHECK(LAG_LOW)
 		if (prob(70))
 			continue
-		if (T && !(T.turf_flags & CAN_BE_SPACE_SAMPLE) && (get_dist(src.get_center(),T) == 4 || get_dist(src.get_center(),T) == 5)) // I'm very tired and this is the least dumb thing I can make of what was here for now
+		if (T && !(T.turf_flags & CAN_BE_SPACE_SAMPLE) && (get_dist(src.get_center(),T) == 4 || get_dist(src.get_center(),T) == 5)) // I'm very tired and this is the least dumb thing I can make of what was here for now.   This needs to get updated for the variable size singularity at some point
```
Comment (ZeWaka) [inline code/obj/machinery/singularity.dm:352]: ?

File: code/obj/machinery/singularity.dm [hunk 22]
```diff
@@ -311,18 +322,18 @@ for some reason I brought it back and tried to clean it up a bit and I regret ev
...
 				return
 		M.take_toxin_damage(3)
-		M.changeStatus("radiation", 100)
+		M.changeStatus("radiation", 20*(radius+1))
 		M.show_text("You feel odd.", "red")
 
 /obj/machinery/the_singularity/proc/Mezzer()
 
-	for (var/mob/living/carbon/M in oviewers(8, src.get_center()))
+	for (var/mob/living/carbon/M in oviewers(radius*4+6, src.get_center()))
```
Comment (ZeWaka) [inline code/obj/machinery/singularity.dm:336]: why *4+6

File: code/obj/machinery/singularity.dm [hunk 23]
```diff
@@ -311,18 +322,18 @@ for some reason I brought it back and tried to clean it up a bit and I regret ev
 
 /obj/machinery/the_singularity/proc/Toxmob()
 
-	for (var/mob/living/carbon/M in orange(7, src.get_center()))
+	for (var/mob/living/carbon/M in orange(radius*3+4, src.get_center()))
```
Comment (ZeWaka) [inline code/obj/machinery/singularity.dm:325]: why *3+4

File: code/obj/machinery/singularity.dm [hunk 24]
```diff
@@ -76,35 +80,40 @@ Contains:
...
 	deconstruct_flags = DECON_WELDER | DECON_MULTITOOL
 
-	bound_width = 96
-	bound_height = 96
-	bound_x = -32
-	bound_y = -32
+	bound_width = 32
+	bound_height = 32
+	bound_x = 16
+	bound_y = 16
```
Comment (ZeWaka) [inline code/obj/machinery/singularity.dm:86]: you might want these to scale with size....

File: code/obj/machinery/singularity.dm (hunk 21 above)
Comment (cogreaper) [reply code/obj/machinery/singularity.dm:352]: my understanding of this was that for a 5x5 singularity it would affect 2 tiles out from the edge of it, I'll fix that now to scale with radius.

File: code/obj/machinery/singularity.dm (hunk 24 above)
Comment (cogreaper) [reply code/obj/machinery/singularity.dm:86]: scaling them created a weird bug that only manifested when the singularity got loose, not having it scale seems to not create any bugs. going to leave it as is until it breaks

**Comment by Azrun** [comment]
I'd like to see radius be something you can "play" with by feeding it a sufficient amount or certain things, but that may be beyond the scope of what you are trying to achieve here.

**Comment by cogreaper** [comment]
> I'd like to see radius be something you can "play" with by feeding it a sufficient amount or certain things, but that may be beyond the scope of what you are trying to achieve here.

The problem with that is that if the radius increases by one, the singularity will immediately free itself as it will grow large enough to cover at least 2 field generators. I considered making it so that when two singularities collide their size would add together, but in a normal setup for a singularity its radius going up by one ends poorly. Maybe its something that could only happen once it gets loose?

File: code/obj/machinery/singularity.dm [hunk 25]
```diff
@@ -21,16 +21,18 @@ Contains:
...
-	var/checkpointC = 0
-	for (var/obj/X in orange(4,src))
-		if (istype(X, /obj/machinery/containment_field))
-			checkpointC ++
-	if (checkpointC >= 20)
+	var/goodgenerators = 0 //ensures that there are 4 generators in place with at least 2 links. note that false positives are very possible and will result in a loose singularity
+	var/smallestdimension = 13//determines the radius of the produced singularity,starts higher than is possible
+
+	for_by_tcl(gen, /obj/machinery/field_generator)//this loop checks for valid field generators
+		if(get_dist(gen,loc)<10)
```
Comment (pali6) [inline code/obj/machinery/singularity.dm:30]: 10 is a bit of a magic number here. Shouldn't this be based off some macro constant that determines the max singularity size?

File: code/obj/machinery/singularity.dm [hunk 26]
```diff
@@ -21,16 +21,18 @@ Contains:
...
+			if(gen.active_dirs>=2)
+				goodgenerators++
+				smallestdimension=min(smallestdimension,gen.shortestlink)
```
Comment (pali6) [inline code/obj/machinery/singularity.dm:33]: Why do you hate whitespace? :(
```suggestion
			if(gen.active_dirs >= 2)
				goodgenerators++
				smallestdimension = min(smallestdimension, gen.shortestlink)
```

File: code/obj/machinery/singularity.dm [hunk 27]
```diff
@@ -21,16 +21,18 @@ Contains:
...
+
+	if (goodgenerators>=4)
```
Comment (pali6) [inline code/obj/machinery/singularity.dm:35]: I'd prefer if this actually did some smart check for singularity containment instead but this is no worse than the old thing so probably ok.

File: code/obj/machinery/singularity.dm (hunk 24 above)
Comment (pali6) [reply code/obj/machinery/singularity.dm:86]: Setting these vars to something else than multiples of 32 makes the game very choppy due to byond internal stuff.

Hoenstly if you are scaling the singulo boundary down to one tile I'd just remove these 4 lines
```suggestion
```
and then update the `get_center()` proc to just return `loc`. As you can see by my earlier exasperated comment in the `get_center()` proc it's all sorta messy. Then you could also remove the ugly `has_moved` var.

File: code/obj/machinery/singularity.dm (hunk 26 above)
Comment (pali6) [reply code/obj/machinery/singularity.dm:33]: Not necessary to get the PR merged but it'd be neat if you used spaces sorta like this in the rest of the PR too pls.

File: code/obj/machinery/singularity.dm [hunk 28]
```diff
@@ -418,8 +434,8 @@ for some reason I brought it back and tried to clean it up a bit and I regret ev
 	var/net_id = null
 	var/obj/machinery/power/data_terminal/link = null
 	mats = 14
-	var/active_dirs = 0
-
+	var/list/active_dirs = 0
```
Comment (pali6) [inline code/obj/machinery/singularity.dm:437]:
```suggestion
	var/active_dirs = 0
```
This is not a list but a bit field.

File: code/obj/machinery/singularity.dm [hunk 29]
```diff
@@ -527,13 +549,17 @@ for some reason I brought it back and tried to clean it up a bit and I regret ev
 	else if(NSEW == 8)
 		oNSEW = 4
 
-	for(var/dist = 0, dist <= 9, dist += 1) // checks out to 8 tiles away for another generator
+	for(var/dist = 0, dist <= 11, dist += 1) // checks out to 11 tiles away for another generator
```
Comment (pali6) [inline code/obj/machinery/singularity.dm:552]: Probably would again benefit from using some sort of a
```
#define MAX_SINGULO_SIZE 11
```
or something similar.

File: code/obj/machinery/singularity.dm [hunk 30]
```diff
@@ -1311,8 +1337,9 @@ for some reason I brought it back and tried to clean it up a bit and I regret ev
 		CAS = locate(/obj/machinery/power/collector_array) in get_step(src,SOUTH)
 		CAE = locate(/obj/machinery/power/collector_array) in get_step(src,EAST)
 		CAW = locate(/obj/machinery/power/collector_array) in get_step(src,WEST)
-		for(var/obj/machinery/the_singularity/S in orange(12,src))
-			S1 = S
+		for_by_tcl(singu, /obj/machinery/the_singularity)//this loop checks for valid singularities
+			if(get_dist(singu,loc)<15)
```
Comment (pali6) [inline code/obj/machinery/singularity.dm:1341]: One more magic number.

**Review by pali6 — CHANGES_REQUESTED** [review]
I love the PR itself. Big singulos look epic.

**Review by ZeWaka — CHANGES_REQUESTED** [review]
see unresolved comments

**Comment by pali6** [comment]
You should probably merge upstream to get rid of the failing checks.

**Comment by pali6** [comment]
Since you removed the `bound_` stuff you should change the `get_center()` proc to just return `src.loc` as it currently tries to adjust for some very wacky byond weirdness with large bounds on objects.

**Review by pali6 — APPROVED** [review]
My comments seem addressed though you should probably address the rest of ZeWaka's comments. (It's mostly just moving stuff to #defines it seems.)

**Comment by cogreaper** [comment]
Somebody pointed out that being able to create an 11x11 singularity instantly would be problematic so I'll modify it so that they start at most as a 5x5 and then grow up to a max radius. I will also likely make them able to grow by eating things once loose.

**Review by ZeWaka — CHANGES_REQUESTED** [review]
unresolved comments

**Review by ZeWaka — APPROVED** [review]
